README.md
# adawat

CHANGELOG.md
---
permalink: /changelog
layout: changelog
---

# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2022-07-21

### Added

-   Added Search page [#123](https://github.com/ARBML/adawat/pull/123) by [@unifoliatus](https://github.com/unifoliatus).
-   Added license [#118](https://github.com/ARBML/adawat/pull/118) by [@zaidalyafeai](https://github.com/zaidalyafeai)
-   Added issue template 
-   Added changelog page [#77](https://github.com/ARBML/adawat/pull/77) by [@sudomaze](https://github.com/sudomaze).
-   Added issue templates for feature request, bug report, and dataset error report [#101](https://github.com/ARBML/adawat/pull/101) by [@MutlaqAldhbuiub](https://github.com/MutlaqAldhbuiub), [#94](https://github.com/ARBML/adawat/pull/94) by [@mazen147](https://github.com/mazen147) and [#92](https://github.com/ARBML/adawat/pull/92) by [@sudomaze](https://github.com/sudomaze).

### Changed

-   Improved stats page [#122](https://github.com/ARBML/adawat/pull/122), [#119](https://github.com/ARBML/adawat/pull/119) by [@zaidalyafeai](https://github.com/zaidalyafeai) and [#114](https://github.com/ARBML/adawat/pull/114) by [@yousef337](https://github.com/yousef337). 
-   Improved the layout template of the website [#81](https://github.com/ARBML/adawat/pull/81) by [@unifoliatus](https://github.com/unifoliatus).
-   Refactoring [#85](https://github.com/ARBML/adawat/pull/85) by [@yousef337](https://github.com/yousef337).
-   Updated `README` with web-service info [#87](https://github.com/ARBML/adawat/pull/87) by [@zaidalyafeai](https://github.com/zaidalyafeai).
-   Improved landing page UI [#89](https://github.com/ARBML/adawat/pull/89) by [@Mosaibah](https://github.com/Mosaibah), [@alonemazin](https://github.com/alonemazin), and [@sudomaze](https://github.com/sudomaze).
-   Improved reporting dataset error [#101](https://github.com/ARBML/adawat/pull/101) by [@MutlaqAldhbuiub](https://github.com/MutlaqAldhbuiub).
-   Improved stats page [#93](https://github.com/ARBML/adawat/pull/93) by [@aalhendi](https://github.com/aalhendi) and [@sudomaze](https://github.com/sudomaze).

## [0.8.0] - 2022-07-11

### Added

-   Added report cards [#56](https://github.com/ARBML/adawat/pull/56) by [@yousef337](https://github.com/yousef337).
-   Added how to run the project locally in `README.md` [#62](https://github.com/ARBML/adawat/pull/62) by [@raed19](https://github.com/raed19).
-   Added `pre-commit` for code-checking and code-formatting [#69](https://github.com/ARBML/adawat/pull/69) by [@sudomaze](https://github.com/sudomaze).

### Changed

-   Updated `contribution guide` and `README` [#69](https://github.com/ARBML/adawat/pull/69) by [@sudomaze](https://github.com/sudomaze).
-   Applied formatting to unify code-formats [#69](https://github.com/ARBML/adawat/pull/69) by [@sudomaze](https://github.com/sudomaze).

### Fixed

-   Fixed `NAN` problem in paper link [#70](https://github.com/ARBML/adawat/pull/70) by [@yousef337](https://github.com/yousef337).
-   Fixed `Collaboration` link at footer [#69](https://github.com/ARBML/adawat/pull/69) by [@sudomaze](https://github.com/sudomaze).

## [0.7.0] - 2022-07-05

### Added

-   Included instructions in `README.md` and `CONTRIBUTING.md` [#48](https://github.com/ARBML/adawat/pull/48) by [@sudomaze](https://github.com/sudomaze).
-   Added a default value for `Graphs` page [#48](https://github.com/ARBML/adawat/pull/48) by [@sudomaze](https://github.com/sudomaze).

### Changed

-   Moved to `jekyll` [#48](https://github.com/ARBML/adawat/pull/48) by [@sudomaze](https://github.com/sudomaze).
-   Updated `CONTRIBUTING.md` [#21](https://github.com/ARBML/adawat/pull/21) by [@unifoliatus](https://github.com/unifoliatus).
-   Refactored JS files for `jekyll` release [#48](https://github.com/ARBML/adawat/pull/48) by [@sudomaze](https://github.com/sudomaze) and [#33](https://github.com/ARBML/adawat/pull/33) by [@yousef337](https://github.com/yousef337).

### Fixed

-   Fixed `cannot load such file...` problem when running `jekyll` [#21](https://github.com/ARBML/adawat/pull/21) by [@unifoliatus](https://github.com/unifoliatus).
-   Fixed `Dialect` page not rendering map [#33](https://github.com/ARBML/adawat/pull/33) by [@yousef337](https://github.com/yousef337).
-   Improved speedup of the website by fetching from `web-service` [#48](https://github.com/ARBML/adawat/pull/48) by [@sudomaze](https://github.com/sudomaze).
-   Updated table content for countries and width [#48](https://github.com/ARBML/adawat/pull/48) by [@sudomaze](https://github.com/sudomaze).

## [0.6.0] - 2021-04-27

### Added

-   Added icon(Favicons) [#6](https://github.com/ARBML/adawat/issues/6) by [@MutlaqAldhbuiub](https://github.com/MutlaqAldhbuiub).
-   Added a new styles [#10](https://github.com/ARBML/adawat/pull/10) by [@alonemazin](https://github.com/alonemazin).

### Changed

-   Refactored the project structure [#7](https://github.com/ARBML/adawat/pull/7) by [@MutlaqAldhbuiub](https://github.com/MutlaqAldhbuiub).
-   Redesigned the pages according to the org. brand [#7](https://github.com/ARBML/adawat/pull/7) by [@MutlaqAldhbuiub](https://github.com/MutlaqAldhbuiub).
-   Refactord `index.js`, re-styled some images & css and made them full-responsive for mobile view [#8](https://github.com/ARBML/adawat/pull/8) by [@alonemazin](https://github.com/alonemazin).
-   Refactored JS/CSS [#10](https://github.com/ARBML/adawat/pull/10) by [@alonemazin](https://github.com/alonemazin).
-   Updated the table and tasks style [#10](https://github.com/ARBML/adawat/pull/10) by [@alonemazin](https://github.com/alonemazin).

### Fixed

-   Updated MaxRowLength anymore [#8](https://github.com/ARBML/adawat/pull/8) by [@alonemazin](https://github.com/alonemazin).

assets/js/card.js
let url = 'https://web-production-a8ed.up.railway.app/datasets/';

function recaptchaChange(){
    let recaptcha_box_checked = (grecaptcha.getResponse()) ? true : false;
    if (recaptcha_box_checked) document.getElementById("grecaptcha-submit").disabled = false;
    else document.getElementById("grecaptcha-submit").disabled = true;
}

function ethicalBadge(text) {
    text = text.toLowerCase();
    if (text == 'low') return '<span class="badge bg-success">Low</span>';
    else if (text == 'medium')
        return '<span class="badge bg-warning">Medium</span>';
    else return '<span class="badge bg-danger text-light">High</span>';
}

function createSubsets(subsetsValue) {
    let result = '<table><tbody>';
    subsetsValue.forEach((subset) => {
        result += `<tr class="border-0"><td class="border-0"><b>${subset["Name"]}</b></td><td class="border-0">${subset["Volume"]}</td></tr>`;
    });
    result += '</tbody></table>';
    return result;
}

function itemize(text) {
    tasks = text.split(',');
    output = '<ul class="list-group list-group-flush bg-transparent">';
    for (let i = 0; i < tasks.length; i++) {
        output +=
            '<li class="list-group-item bg-transparent p-0">' +
            tasks[i].trim().replaceAll(' ', '-') +
            '</li>';
    }
    output += '</ul>';
    return output;
}

function injectPaperName(paperName) {
    // inject paper name
    paperNameSpans = document.getElementsByClassName('paperNameSpan');
    // change the values of the spans without loop
    paperNameSpans[0].innerText = paperName;
    paperNameSpans[1].innerText = paperName;
    // append the paper name to href of the link 
    paperNameLink = document.getElementById('paperATag');
    paperNameLink.href = `${paperNameLink.href}${paperName}`;
}

async function onSendReportBtnClicked() {
    let issueForm = document.getElementById('issueMessage');
    const cardId = new URL(window.location.href).searchParams.get('id');

    const response = await fetch(`${url}${cardId}/issues`, {
        method: 'POST',
        body: JSON.stringify({ body: issueForm.value }),
        headers: { 'Content-Type': 'application/json' },
    });

    if (response.ok) {
        const responseData = await response.json();

        issueForm.value = '';
        tata.success(
            'Issues opend succesfully',
            `Track at ${responseData['issue_url']}`,
            {duration: 15000}
        );
    } else {
        tata.error(
            'Error during opening the issue',
            'Please try again later, or contact us via our discord server'
        );
    }
}

// get id from page parameters
const urlParams = new URLSearchParams(window.location.search);
const card_id = urlParams.get('id');

axios
    .get(url + card_id)
    .then(function (response) {
        let headers = [];
        let headersWhiteList = [
            "Id","Name","Link","Colab link","GitHub Repo","Pricing","Accessibility","License","Version","Description","Paper Title","Paper URL","Release Year","Tasks","Supported language(s)","Tool Type","Interface","Programming Language","Evaluated datasets","Cluster","Embeddings"
        ];

        $('.loading-spinner').hide();
        for (let i = 0; i < headersWhiteList.length; i++) {
            headers.push({
                index: i,
                title: headersWhiteList[i],
            });
        }

        let row = response.data;
        injectPaperName(row.Name);

        let dataset = [];

        for (let i = 0; i < headers.length; i++) {
            let element = headers[i];
            let value =
                row[element.title] != 'nan' ? row[element.title] : 'N/A';
            // console.log(element.title, value);
            if (value == 'N/A') {
                dataset.push({
                    0: element.title,
                    1: '',
                });
                continue;
            }
            if (element.title == 'Ethical Risks') {
                value = ethicalBadge(value); // calling "ethicalBadge" function to put some style to the value
            } else if (
                element.title == 'Link' ||
                element.title == 'Paper URL' ||
                element.title == 'Colab link'
            ) {
                value = linkuize(value, value);
            } else if (element.title == 'Subsets') {
                if (row[element.title] != 'nan') {
                    let subsets = row[element.title];
                    value = createSubsets(subsets);
                }
            } else if (element.title == 'Tasks') {
                value = itemize(value);
            }
            dataset.push({
                0: element.title,
                1: value,
            });
        }
        // console.log(dataset);
        $(document).ready(function () {
            $('#table_card').DataTable({
                data: dataset,
                columns: [
                    {
                        title: 'Attribute',
                    },
                    {
                        title: 'Value',
                    },
                ],
                lengthMenu: [10, 25, 50, 75, 100, 250],
                scrollCollapse: true,
                // scrollY: "720px",
                scrollCollapse: true,
                paging: false,
                order: [],
                bInfo: false,
                searching: false,
            });
        });
    })
    .catch(function (error) {
        console.log(error);
    });

document
    .getElementById('grecaptcha-submit')
    .addEventListener('click', onSendReportBtnClicked);


assets/js/index.js
const url = 'https://web-production-a8ed.up.railway.app/datasets';
const contributers_url = 'https://web-production-a8ed.up.railway.app/datasets/tags?features=Added by'

function linkuize(text, link, short = true) {
    if (link != undefined && link != 'nan')
        return `<a href = "${link}" target="_blank" class="#${(short) ? "shorterText " : ""}underline"> ${text}</a>`;
    else return 'Not Available';
}

function getIcon(text) {
    const lower = text.toLowerCase();
    if (icons[lower] != undefined || icons[lower] != 'nan') {
        return icons[lower];
    } else {
        return text;
    }
}
function itemize(text) {
    tasks = text.split(',');
    output = '<ul class="list-group list-group-flush bg-transparent">';
    for (let i = 0; i < tasks.length; i++) {
        output +=
            '<li class="list-group-item bg-transparent">' +
            tasks[i].trim().replaceAll(' ', '-') +
            '</li>';
    }
    output += '</ul>';
    return output;
}

function createColab(link){
    return `<a href=${link}>
    <img src="https://colab.research.google.com/assets/colab-badge.svg" width = '400px' >
  </a>`
}
function badgeRender(text) {
    text = text.toString().toLowerCase();
    if (text.toLowerCase() == 'free')
        return '<span class="text-sm font-medium  px-2.5 py-0.5" style="background-color: #00800030; color:green; font-weight:bold; border-radius:5px">Free</span>';
    else if (text == 'upon-request')
        return '<span class="badge bg-info px-2.5 py-2">Free Upon Request</span>';
    else return '<span class="badge bg-danger px-2.5 py-2">Paid</span>';
}

function reformat_numbers(num) {
    values = num.split(',');
    if (values.length < 2) {
        return num;
    } else if (values.length == 2) {
        return values[0] + 'K';
    } else return values[0] + 'M';
}

function reformat_version(num) {
    values = num.split('.');
    if (num == "nan")
        return ""
    if (values.length == 2) {
        return num+".0";
    } else if (values.length == 1) {
        return num+".0.0";
    } else return num;
}

async function getDetails(id) {
  return axios.get(url+"/"+id).then(response => response.data)
}

async function getOGimage(url) {
    if (url.includes("github"))
    {
        let owner = url.split("/")
        let preview = `https://opengraph.githubassets.com/1/${owner[3]}/${owner[4]}`
        return preview
    }
    else
    return "./assets/images/logo.png"
}

async function fomratDetails(data, index){
    // console.log(data, "s")
    await getOGimage(data['Link']).then(res => {
        return (res) ? image = res : image = "./assets/images/logo.png"

    })
  return '<div class="grid grid-cols-4">'+
            '<div class="col-span-1">'+
                // '<a class="text-center fs-3">'+ linkuize(data['Paper Title'], data['Paper Link'])+'</a>'+
                // '<a href = "'+data['Link']+'" target="_blank" class="shorterText underline mx-4" style="width: 70%"> '+data['Link']+'</a>'+
                '<a style="line-height: 9rem;" target="_blank" href="' + data['Link'] + '"><img style="width: 70%;" class="shorterText underline mx-4" src="'+ image +'"/></a>'+

            '</div>'+
            '<div class="col-span-3 relative ">'+
                '<div class="grid grid-rows-6 grid-flow-col ">'+
                    ' <div class="grid grid-cols-2 ">'+
                        '<span class="text-gray-400">Name</span>'+
                        '<span class="text-gray-800">'+data['Name'] +'</span>'+
                    '</div>'+
                    ' <div class="grid grid-cols-2">'+
                        '<span class=" text-gray-400">Created At</span>'+
                        '<span class=" text-gray-800">'+data['Release Year'] +'</span>'+
                    '</div>'+
                    ' <div class="grid grid-cols-2 ">'+
                        '<span class="text-gray-400">Version</span>'+
                        '<span class="text-gray-800">'+data['Version'] +'</span>'+
                    '</div>'+
                    ' <div class="grid grid-cols-2  ">'+
                        '<span class="text-gray-400">Accessibility</span>'+
                        '<span class="text-gray-800">'+data['UnAccessibilityit'] +'</span>'+
                    '</div>'+
                    ' <div class="grid grid-cols-2  ">'+
                        '<span class="text-gray-400">Supported language(s)</span>'+
                        '<span class="text-gray-800">'+data['Supported language(s)'] +'</span>'+
                    '</div>'+
                    ' <div class=" grid grid-cols-2 ">'+
                        '<span class="text-gray-400">License</span>'+
                        '<span class="text-gray-800">'+data['License'] +'</span>'+
                    '</div>'+
                    ' <div class="grid grid-cols-2  ">'+
                        '<span class="text-gray-400">Accessibility</span>'+
                        '<span class="text-gray-800">'+data['Accessibility'] +'</span>'+
                    '</div>'+
                    ' <div class="grid grid-cols-2  ">'+
                        '<span class="text-gray-400">Tasks</span>'+
                        '<span class="text-gray-800">'+data['Tasks'] +'</span>'+
                    '</div>'+
                    ' <div class="grid grid-cols-2  ">'+
                        '<span class="text-gray-400">Language</span>'+
                        '<span class="text-gray-800">'+data['Programming Language'] +'</span>'+
                    '</div>'+
                    ' <div class="grid grid-cols-2  ">'+
                        '<span class="text-gray-400">GitHub Repo</span>'+
                        '<span class="text-gray-800">'+data['GitHub Repo'] +'</span>'+
                    '</div>'+
                    ' <div class="grid grid-cols-2  ">'+
                        '<span class="text-gray-400">Pricing</span>'+
                        '<span class="text-gray-800">'+data['Pricing'] +'</span>'+
                    '</div>'+
                    ' <div class="grid grid-cols-2  ">'+
                        '<span class="text-gray-400">Tool Type</span>'+
                        '<span class="text-gray-800">'+data['Tool Type'] +'</span>'+
                    '</div>'+
                    ' <div class=" grid grid-cols-2 ">'+
                        '<span class="text-gray-400">Interface</span>'+
                        '<span class="text-gray-800">'+data['Interface'] +'</span>'+
                    '</div>'+
                    ' <div class="grid grid-cols-2  ">'+
                        '<span class="text-gray-400">Evaluated datasets</span>'+
                        '<span class="text-gray-800">'+data['Evaluated datasets'] +'</span>'+
                    '</div>'+
                '</div>'+
                '<div class="collapse-footer flex justify-end gap-x-5 mt-7">'+
                '<a href="'+`card?id=${index}`+ '" class="underline font-normal">Details</a>'+
                '<a href="'+data["Paper Link"]+'" target="_blank" class="underline font-normal">Paper</a>'+
            '</div>'+
        '</div>'
   // })
 }

 async function getContributersNum()
 {
    try {
        let res = await axios({
             url: contributers_url,
             method: 'get',
         }) 
         return res.data["Added by"].length
     }
     catch (err) {
         console.error(err);
     }
}

axios
    .get(url, {
        onDownloadProgress: (progressEvent) => {},
    })
    .then(function (response) {
        let headers = [];
        let headersWhiteList = [
            'No.',
            'Name',
            'Colab Link',
            'Year',
            'License',
            'Version',
            'Language',
            'Paper Link',
            'Access',
            'Tasks',
        ];
        $('.loading-spinner').hide();
        headers.push({
          index: 0,
          className:      'fa-solid button table-cell flex rounded-tl-lg',
          orderable:      false,
          data:           null,
          defaultContent: ''
      
      })
        for (let i = 0; i < headersWhiteList.length; i++) {
            headers.push({
                index: 1+i,
                title: headersWhiteList[i].toUpperCase(),
            });
        }
        // console.log(headers[headers.length - 1].className = "rounded-tr-lg ")
        let rows = response.data;
        // console.log(headers);

        //  Createing table data
        let dataset = [];
        for (let index = 0; index < rows.length; index++) {
            const row = rows[index];
            var host = row['Host']
            if (host == 'other') host = "External Link"
            let link_host = linkuize(host, row['Link']);
            if (row['HF Link'] != 'nan') {
                link_host += '</br>' + linkuize(getIcon('hf'), row['HF Link']);
            }
            console.log(row["Version"])
            dataset.push({
                0: index + 1,
                1: index + 1,
                2: linkuize(row['Name'], `card.html?id=${row['Id']}`, false),
                3: createColab(row["Colab link"]),
                4: parseInt(row['Release Year']),
                5: row["License"],
                6: reformat_version(row["Version"]),
                7: row["Programming Language"],
                8: linkuize(row['Paper Title'], row['Paper URL']),
                9: badgeRender(row['Pricing']),
                10: itemize(row['Tasks']),
            });
        }

        $.extend($.fn.dataTableExt.oSort, {
            'data-custom-pre': function (a) {
                // console.log(a);
            },
        });

        $(document).ready(function () {
            document.getElementById('numDatasets').textContent = dataset.length;
            getContributersNum().then(res => 
                document.getElementById('numContributers').textContent = res)
            let table = $('#table').DataTable({
                data: dataset,
                columns: headers,
                lengthMenu: [
                    [10, 100, 200, 300, 400, -1],
                    [10, 100, 200, 300, 400, 'All'],
                ],
                scrollCollapse: true,
                paging: true,
                pagingType: 'numbers',
                bInfo: false,
                dom: 'pt',
                createdRow: function (row, data, dataIndex) {
                    $('td:eq(`10`)', row).css('min-width', '200px');
                },
                order: [[1, 'asc']],
                searching: true
              });
          
             // opening and closing details
             $('#table tbody').on('click', 'td.button', function () {
                var tr = $(this).closest('tr');
                var row = table.row( tr );
                if ( row.child.isShown() ) {
                    row.child.hide();
                    tr.removeClass('shown');
                }
                else {
                    id = row.data()[1];
                    loader = $(".loading-spinner").html();
                    row.child(loader).show();
                    getDetails(id).then(async (response) =>
                      row.child(await fomratDetails(response, id)).show()
                    );
                    tr.addClass('shown');
                }
              });
        });
    })
    .catch(function (error) {
        console.log(error);
    });


assets/js/search.js
const request = axios.create({
  baseURL: "https://web-production-a8ed.up.railway.app",
});

function ethicalBadge(text) {
  text = text.toLowerCase();
  if (text == 'free') return '<span class="badge bg-success">Free</span>';
  else return '<span class="badge bg-danger text-light">Paid</span>';
}

function accessBadge(text) {
  text = text.toString().toLowerCase();
  if (text.toLowerCase() == 'free')
      return '<span class="text-sm font-medium  px-2.5 py-0.5" style="background-color: #00800030; color:green; font-weight:bold; border-radius:5px">Free</span>';
  else if (text == 'upon-request')
      return '<span class="badge bg-info px-2.5 py-2">Free Upon Request</span>';
  else return '<span class="badge bg-danger px-2.5 py-2">Paid</span>';
}

function setAttributes(attribute, element){

  if (attribute == "Cost"){
    if(element[attribute] != "nan") return element[attribute]
    else return "0$"
  }else if (attribute == "Access"){
    return accessBadge(element[attribute])
  }
  else if (element[attribute] != "nan"){
    return element[attribute]
  }

  return "unknown"
  
}

const shove = {
  Tasks: [
    "machine translation",
    "speech recognition",
    "information retrieval",
    "sentiment analysis",
    "language modeling",
    "topic classification",
    "dialect identification",
    "text generation",
    "cross-lingual information retrieval",
    "named entity recognition",
    "question answering",
    "information detection",
    "summarization",
    "speaker identification",
    "hate speech detection",
    "morphological analysis",
    "translation",
    "information extraction",
  ],
};

const entries = [
  "Id","Name","Link","Colab link","GitHub Repo","Pricing","Accessibility","License","Version","Description","Paper Title","Paper URL","Release Year","Tasks","Supported language(s)","Tool Type","Interface","Programming Language","Evaluated datasets"
];

const queries = new URLSearchParams(window.location.search);

/**
 * this should return boolean of whatever the key are provided through query or not.
 * @param {string} k
 * @returns {boolean}
 */
const isProvided = (k) =>
  queries.has(k) && (queries.get(k) > 0 || queries.get(k).length > 0);

/**
 * @type {Set<HTMLInputElement>}
 */
const listOfToggable = new Set();

/**
 * @type {HTMLLabelElement}
 */
const form = document.querySelector("#form");

if (isProvided("name")) {
  const name = queries.get("name");
  $("#special").text(name);
  $("input[name='name']").each((_, e) => $(e).val(name));
} else {
  $("#special").text("not provided".toUpperCase());
}

form.addEventListener("submit", (event) => {
  event.preventDefault();

  $("#loader").removeClass("hidden");

  const name = $("#form input[name='name']").val();

  if (!name) queries.delete("name");
  else {
    queries.set("name", $("#form input[name='name']").val());
  }

  $("input[name='name']").each((_, e) => $(e).val(`${name}`));

  $("#special").text(name || "not provided".toUpperCase());

  const url = new URL(window.location);

  url.search = `${queries}`;

  window.history.pushState({}, "", url);

  $("#pagination").pagination({
    ulClassName: "pagination",
    dataSource: (cb) => {
      const parameter = new URLSearchParams({
        query: [
          ...(isProvided("name")
            ? [`Name.str.contains('(?i)${queries.get("name")}')`]
            : []),
          ...(isProvided("license") ? [`License.str.contains('(?i)${queries.get("license")}')`] : []),
          ...(isProvided("pricing") ? [`Pricing.str.contains('(?i)${queries.get("pricing")}')`] : []),
          ...(isProvided("since") ? [`Year > ${queries.get("since")}`] : []),
          ...(isProvided("afore") ? [`Year < ${queries.get("afore")}`] : []),
          ...(listOfToggable.size > 0
            ? Array.from(listOfToggable.values())
                .filter((e) => $(e).children("input")[0].checked)
                .map((e) => $(e).children("input")[0])
                .map((e) => `${e.name}.str.contains('${e.value}')`)
            : []),
        ].join(" and "),
        features: entries,
      });
      request
        .get(`/datasets?${parameter}`)
        .then((response) => response.data)
        .then(cb);
    },
    callback: (data) => {
      let html = "";

      for (let element of data) {
        html += "<li class='flex flex-col gap-3' >";
        html += `<a class='font-bold text-primary' target='_blank' href="${element["Link"]}">${element["Name"]}</a>`;

        html +=
          "<div class='grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-x-12 gap-y-2'>";
        
        html += "<div class='flex justify-between gap-3'>";
        html += `<span class='font-bold capitalize text-gray-600 whitespace-nowrap'>ID</span>`;
        html += `<span class='truncate'>${element.Id}</span>`;
          html += "</div>";
        for (let attribute in element) {
          if (["Id", "Pricing", "Paper URL", "Link", "Name"].includes(attribute))
            continue;
          html += "<div class='flex justify-between gap-3'>";
          html += `<span class='font-bold capitalize text-gray-600 whitespace-nowrap'>${attribute}</span>`;
          html += `<span class='truncate'>${
            setAttributes(attribute, element)
          }</span>`;
          html += "</div>";
        }
        html += "</div>";

        html += "<div class='flex justify-between'>";
        html += `<span class='capitalize font-bold'> Pricing ${ethicalBadge(element["Pricing"])} </span>`;
        html += "<div/>";

        html += "<div class='gap-2 flex'>";
        html += `<a class='capitalize font-bold text-primary' target='_blank' href='./card?id=${element["Id"]}'>details</a>`;
        html += `<a class='capitalize font-bold text-primary' target='_blank' href='${element["Paper URL"]}'>paper link</a>`;
        html += "</div>";

        html += "</li>";
        html += "<li><hr/></li>";
      }

      $("#list-of-detail").html(html);

      $("#loader").addClass("hidden");
    },
  });
});

(() => {
  const queries = new URLSearchParams();

  queries.set("features", ["Tasks", "Pricing", "License"]);

  request
    .get(`datasets/tags?${queries}`)
    .then((response) => response.data)
    .then((object) => {
      for (let k in object) {
        document.querySelector(`#${k} > div`).addEventListener("click", (e) => {
          $(`#${k} > .body`).toggleClass("max-h-0");
          $(`#${k} > div > i`).toggleClass("fa-chevron-up");
        });

        if (shove[k])
          object[k] = object[k].sort((e) => (shove[k].includes(e) ? -1 : 1));

        for (let index in object[k]) {
          const value = object[k][index];
          const element = document.createElement("label");

          element.classList.add(
            "bg-black/[.02]",
            "border-black/50",
            "text-black/50",
            "border-1",
            "cursor-pointer",
            "rounded-full",
            "py-0.5",
            "px-2.5"
          );

          if (index > 7) element.classList.add("hidden");

          {
            const el = document.createElement("input");

            el.name = k;

            el.value = value;

            el.type = "checkbox";

            el.classList.add("invisible", "w-0", "h-0");

            el.addEventListener("change", (event) => {
              [
                "bg-black/[.02]",
                "border-black/50",
                "text-black/50",
                "bg-[#be2596]/50",
                "border-[#be2596]",
                "text-[#be2596]",
              ].forEach((className) => element.classList.toggle(className));
              submit();
            });

            element.appendChild(el);
          }

          {
            const el = document.createElement("span");

            el.textContent = value;

            element.appendChild(el);
          }

          listOfToggable.add(element);
          document
            .querySelector(`#${k} > .body > ul.options`)
            .appendChild(element);
        }

        const elements = Array.from(listOfToggable).filter(
          (element) => $(element).children("input")[0].name == k
        );

        /**
         * @type {HTMLInputElement}
         */
        const expander = document.querySelector(`#${k} input.expander`);

        /**
         * @type {HTMLInputElement}
         */
        const through = document.querySelector(`#${k} input.through`);

        through.addEventListener("keyup", (event) => {
          const value = event.target.value.toLowerCase();

          if (value.length > 0) $(expander).parent().addClass("hidden");
          else $(expander).parent().removeClass("hidden");

          elements.forEach((el, index) => {
            const isMatching = $(el)
              .children("input")[0]
              .value.toLowerCase()
              .includes(value);

            if ((value.length <= 0 && index > 7) || !isMatching)
              return el.classList.add("hidden");

            el.classList.remove("hidden");
          });
        });

        expander.addEventListener("change", (event) => {
          /**
           * @type {HTMLInputElement}
           */
          const element = event.target;

          /**
           * @type {boolean}
           */
          const isChecked = element.checked;

          elements.map((el, index) => {
            if (!isChecked && index > 7) return el.classList.add("hidden");
            el.classList.remove("hidden");
          });

          $(element)
            .siblings("span")
            .text(`show ${isChecked ? "less" : "more"}`);
        });
      }
    });
})();

document.querySelector("#reset").addEventListener("click", () => {
  Array.from(listOfToggable).map((element) => {
    const isChecked = $(element).children("input")[0].checked;

    if (isChecked)
      element.classList.remove(
        "bg-[#be2596]/50",
        "border-[#be2596]",
        "text-[#be2596]"
      );

    element.classList.add("bg-black/[.02]", "border-black/50", "text-black/50");

    $(element).children("input")[0].checked = false;
    return false;
  });

  Array.from(document.querySelectorAll("#Year input").values()).map(
    (e) => (e.value = "")
  );

  queries.delete("afore");
  queries.delete("since");
  submit();
});

document.querySelectorAll("#Year input").forEach((e) => {
  // ! for some reason keyup mess up the layout because of many request to the form. ig
  e.addEventListener("blur", (event) => {
    queries.set(event.target.name, event.target.value);
    submit();
  });
});

document.querySelector(`#Year > div`).addEventListener("click", (e) => {
  $(`#Year > .body`).toggleClass("max-h-0");
  $(`#Year > div > i`).toggleClass("fa-chevron-up");
});

function submit() {
  $("#submit").click();
}

submit();
